text/0000-otlp-trace-data-format.md

```diff
@@ -0,0 +1,383 @@
+# OTLP Trace Data Format
+
+_Author: Tigran Najaryan, Omnition Inc._
+
+**Status:** `proposed`
+
+OTLP Trace Data Format specification describes the structure of the trace data that is transported by OpenTelemetry Protocol (RFCNNN).
+
+## Motivation
+
+This document is a continuation of OpenTelemetry Protocol RFCNNN and is necessary part of OTLP specification.
+
+## Explanation
+
+OTLP Trace Data Format is primarily inherited from OpenCensus protocol. Several changes are introduced with the goal of more efficient serialization. Notable differences from OpenCensus protocol are:
+
+1. Removed `Node` as a concept.
+2. Extended `Resource` to better describe the source of the telemetry data.
+3. Replaced attribute maps by lists of key/value pairs.
+4. Eliminated unnecessary additional nesting in various values.
+
+Changes 1-2 are conceptual, changes 3-4 improve performance.
+
+## Internal details
+
+This section specifies data format in Protocol Buffers.
+
+### Resource
+
+```
+// Resource information. This describes the source of telemetry data.
+message Resource {
+  // Identifier that uniquely identifies a process within a VM/container.
+  Process process = 1;
+
+  // Name of the service.
+  string service_name = 2;
+
+  // Set of labels that describe the resource.
+  repeated AttributeKeyValue attributes = 3;
+}
+
+// Identifies a process within a VM/container.
+message Process {
```

Comment:
Why is this important?

Reply:
>we need a notion of instance_id in that service (when service has more than one binary instance running).

I agree, this is very important, I wanted to discuss this with you too.

Reply:
I think you responded to a different comment and did not address my question about `Process`

Reply:
Sorry, you are right, I replied to a different comment.

To answer the question about `Process`. This is derived from OpenCensus Proto which has the concept of `ProcessInformation` in the Node.

Basically I believe we need to record information about the origin of the trace. I think process id and hostname are important for identification on the backend that the spans are generated from the same application instance.

Another approach (and what you mentioned in another comment) is to have a notion of unique instance_id, which will be different for each instance of a service (even same service). This can be a generated UUID or something else, the only requirement is that it must be unique per instance of a service.

I think this is important because it will allow to correlate data coming from the same source but probably traveling via different paths, traveling at different times, or being of different type. Imagine for example that as end user I am examining a particular span in the backend and I want to also see corresponding infrastructure metrics for the machine (or container) where this span is coming from. Having a unique ID of instance present in all telemetry data (both traces and metrics) will help to do it.

I am not sure if `Process` containing process id and hostname is the right approach. What is your opinion on this?

Reply:
I think "Resource" should uniquely identify the source of the telemetry:
* Cluster information
* Host
* Container (if available)
* Service (including service name and instance in that service)

So unless we need explicitly the hostname/pid I would not add them (hostname is already available in the Host). See https://github.com/census-instrumentation/opencensus-specs/blob/master/resource/StandardResources.md

Reply:
One important thing that Resource does not do is it may not allow you to communicate with the source (some kind of ip:port where you can for example push config or other things).

Reply:
Sounds good. I am going to remove these fields from this data format and I will submit a separate PR to add Standard Resource labels to the specification. I will take OpenCensus standard as the basis and will add "service" labels.

Reply:
Fields are removed from this PR. Preparing a separate PR to add standard resource labels to the specification repo.